File: testdata/corpus/py/greeter/store.py
"""Reaching the corpus schema from Python. Corpus fixture: not installed, not run.

The three shapes go/store/store.go's doc comment describes, in Python's spellings: a
triple-quoted literal, an f-string whose table is a parameter, and a log message.
"""

LIST_ORDERS = """
SELECT id, total
FROM orders
WHERE total > %s
"""


def list_orders(cur):
    """Reads one table."""
    cur.execute(LIST_ORDERS, (0,))
    return cur.fetchall()


def record(cur):
    """Writes the table it names."""
    cur.execute("INSERT INTO customers (id) VALUES (%s)", (1,))


def purge(cur, table):
    """The gap. An f-string's `{table}` is not a name in the tree."""
    cur.execute(f"DELETE FROM {table} WHERE total = 0")


def warn(table):
    """Prose. `.format` on a sentence is not a statement with an interpolated name."""
    return "delete from {} returned no rows".format(table)

File: testdata/corpus/ts/packages/api/src/store.ts
// Reaching the corpus schema from TypeScript. Corpus fixture: not built, not run.
//
// The three shapes go/store/store.go's doc comment describes, in TypeScript's spellings. The
// template literal is the interesting one: it is both how a multi-line query is written and how
// a name is interpolated, so the two boundaries share one delimiter here and the reader has to
// tell them apart by what is inside it rather than by which quote opened it.

const LIST_ORDERS = `
SELECT id, total
FROM orders
WHERE total > $1
`;

export async function listOrders(db: { query(sql: string): Promise<unknown> }) {
  return db.query(LIST_ORDERS);
}

export async function record(db: { query(sql: string): Promise<unknown> }) {
  return db.query("INSERT INTO customers (id) VALUES ($1)");
}

// The gap: same delimiter, and the table is a parameter.
export async function purge(
  db: { query(sql: string): Promise<unknown> },
  table: string,
) {
  return db.query(`DELETE FROM ${table} WHERE total = 0`);
}

// Prose.
export function warn(table: string): string {
  return `insert into ${table} failed, retrying`;
}
